nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchHttpRecord.java

```diff
@@ -0,0 +1,559 @@
+                    break;
+            }
+        }
+        return problems;
+    }
+
+    @OnScheduled
+    public void setup(ProcessContext context) {
+        super.setup(context);
+        recordPathCache = new RecordPathCache(10);
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final RecordReaderFactory readerFactory = context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+
+        // Authentication
+        final String username = context.getProperty(USERNAME).evaluateAttributeExpressions(flowFile).getValue();
+        final String password = context.getProperty(PASSWORD).evaluateAttributeExpressions(flowFile).getValue();
+
+        OkHttpClient okHttpClient = getClient();
+        final ComponentLog logger = getLogger();
+
+        final String baseUrl = trimToEmpty(context.getProperty(ES_URL).evaluateAttributeExpressions().getValue());
+        final URL url;
+        try {
+            url = new URL((baseUrl.endsWith("/") ? baseUrl : baseUrl + "/") + "_bulk");
+        } catch (MalformedURLException mue) {
+            // Since we have a URL validator, something has gone very wrong, throw a ProcessException
+            context.yield();
+            throw new ProcessException(mue);
+        }
+
+        final String index = context.getProperty(INDEX).evaluateAttributeExpressions(flowFile).getValue();
+        if (StringUtils.isEmpty(index)) {
+            logger.error("No value for index in for {}, transferring to failure", new Object[]{flowFile});
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+        }
+        final String docType = context.getProperty(TYPE).evaluateAttributeExpressions(flowFile).getValue();
+        String indexOp = context.getProperty(INDEX_OP).evaluateAttributeExpressions(flowFile).getValue();
+        if (StringUtils.isEmpty(indexOp)) {
+            logger.error("No Index operation specified for {}, transferring to failure.", new Object[]{flowFile});
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+        }
+
+        switch (indexOp.toLowerCase()) {
+            case "index":
+            case "update":
+            case "upsert":
+            case "delete":
+                break;
+            default:
+                logger.error("Index operation {} not supported for {}, transferring to failure.", new Object[]{indexOp, flowFile});
+                session.transfer(flowFile, REL_FAILURE);
+                return;
+        }
+
+        final String id_path = context.getProperty(ID_RECORD_PATH).evaluateAttributeExpressions(flowFile).getValue();
+        final RecordPath recordPath = StringUtils.isEmpty(id_path) ? null : recordPathCache.getCompiled(id_path);
+        final StringBuilder sb = new StringBuilder();
```

Comment:
Since we are going to be building up this big string of commands, should we document in the description of the processor that all records in the incoming flow file will be read into memory, in case someone sends in flow file with millions of records?